src/twin/ssh_tunnel.rs

```diff
@@ -0,0 +1,449 @@
+}
+
+fn get_pub_key_path() -> PathBuf {
+    Path::new(SSH_KEY_DIR).join(format!("{}.pub", SSH_KEY_NAME))
+}
+
+fn get_priv_key_path() -> PathBuf {
+    Path::new(SSH_KEY_DIR).join(SSH_KEY_NAME)
+}
+
+fn get_cert_path(name: &str) -> PathBuf {
+    Path::new(SSH_KEY_DIR).join(format!("{}-{}-cert.pub", SSH_KEY_NAME, name))
+}
+
+fn get_control_socket(name: &str) -> PathBuf {
+    Path::new(&*CONTROL_SOCKET_DIR).join(name)
+}
+
+fn create_control_socket_path() -> Result<()> {
+    Ok(fs::create_dir_all(&*CONTROL_SOCKET_DIR)?)
+}
+
+fn create_or_get_pub_key() -> Result<Vec<u8>> {
+    // lock against concurrent access to the key pair.
+    static KEY_FILE_LOCK: Lazy<Mutex<()>> = Lazy::new(|| Mutex::new(()));
+    let _guard = KEY_FILE_LOCK.lock().unwrap();
+
+    let pub_key_path = get_pub_key_path();
+
+    // check if there is already a key pair, if not, create it first
+    if !pub_key_path.exists() {
+        fs::create_dir_all(SSH_KEY_DIR).unwrap();
+        let result = std::process::Command::new("ssh-keygen")
+            .stdout(Stdio::piped())
+            .args(["-q"])
+            .args(["-f", get_priv_key_path().to_str().unwrap()])
+            .args(["-t", SSH_KEY_TYPE])
+            .args(["-N", ""])
+            .output();
+
+        match &result {
+            Err(error) => {
+                error!("Failed to create ssh key pair: {}", error);
+                anyhow::bail!("Error on ssh key creation.");
+            }
+            Ok(output) => {
+                if !output.status.success() {
+                    error!(
+                        "Failed to create ssh key pair: {}",
+                        str::from_utf8(&output.stderr).unwrap()
+                    );
+                    anyhow::bail!("Error on ssh key creation.");
+                }
+            }
+        }
+    }
+
+    match fs::read(pub_key_path.to_str().unwrap()) {
+        Ok(pub_key) => Ok(pub_key),
+        Err(err) => {
+            error!("Failed to open ssh public key: {}", err);
+            anyhow::bail!("Error reading ssh public key.");
+        }
+    }
+}
+
+// RAII handle to the temporary bastion host certificate
+struct CertificateFile {
+    path: PathBuf,
+}
+
+impl CertificateFile {
+    fn new(path: impl AsRef<Path>, data: &[u8]) -> Result<CertificateFile> {
+        let mut f = fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .mode(0o600)
+            .open(path.as_ref())?;
+        f.write_all(data)?;
+        Ok(CertificateFile {
+            path: path.as_ref().to_path_buf(),
+        })
+    }
+}
+
+impl Drop for CertificateFile {
+    fn drop(&mut self) {
+        if let Err(err) = fs::remove_file(&self.path) {
+            warn!(
+                "Failed to delete certificate \"{}\": {}",
+                self.path.to_str().unwrap(),
+                err
+            );
+        }
+    }
+}
```

Comment:
just for my understanding: we also delete the file whenever the service shuts down? what happens if we have a crash? are we able to recover?